### ISSUE TITLE
We don’t care about SNI reliability, server certificate is indeed reliable ! 

### ISSUE THREAD

**Issue by zied-turki** [echdeploy/draft-ech-deployment-considerations#65]
Hi all

**TLDR;** 
I think that our opponents have identified the weakest link on the current text (regarding the use of SNI by the middleboxes in enterprises) which is indeed the best angle of attack for them.
Let’s take a step back and ask ourselves how do middleboxes do to identify the service requested by the user to perform web categorization to do selective inspection or content/web filtering?
Actually, middleboxes don’t use SNI to perform selective inspection but rather use server certificate information. SNI is only used to fetch the destination server certificate with TLS 1.3. Web categorization is only done based on the server certificate details but not the SNI..

Please share your views and thoughts.

### **Detailed explanation**
I believe that we are all aligned that the SNI is by design not reliable. Btw, referring to RFC 6066, TLS extensions, including SNI, are designed to be backward compatible. This means that if the server doesn’t recognize the SNI value, the TLS handshake should continue anyway. 
In other terms, SNI value can 1) be empty or 2) have an alternative name which is different from the real name of the destination server without impacting the establishment of the TLS session. 
This can also be easily exploited by bad actors indeed to bypass security middle boxes. E.g A malware could just be coded to provide an SNI value that is mapped to finance or healthcare categories to bypass inspection (it is not that easy, but there is a way to do it).
Here below my understanding of how it works for most of proxy servers in the market. Please share your views / thoughts.

**With TLS 1.2**
User client generates “ClientHello” with SNI in plain text to the destination server. If accepted, the server responds to the user client request with the “ServerHello” message containing the intended server certificate alongside other encryption-related information in the plain text mode as well. 
Middleboxes can then see/inspect 1/ the SNI in the “ClientHello” and 2/ the server certificate details in the “ServerHello” message.
Middleboxes can perform selective inspection based on the destination service details (service requested by the user client) extracted from the server certificate. 
I would like to note that depending on the middlebox design, it is more about a correlation of different source of information to confirm the right destination service. But still only information within the server certificate is reliable to perform accurate web categorization and then do selective inspection of any kind of web/content filtering. 

**With TLS 1.3**
TLS 1.3 improves significantly over the TLS 1.2 in terms of security and privacy aspects. 
More specifically, in terms of privacy, it overcomes the plain text server certificate exchange issue by masking the server’s host identity through the encrypted server certificate. So as middleboxes inspection capabilities are designed based on the server certificate, all vendors worked on how to adapt their capabilities to support TLS 1.3 (e.g ServerHello encryption but not only).
So how do providers support TLS 1.3 inspection then?
The most common technique is first to get the SNI from the ClientHello (which is still shared in plain text format.) and then replays / establish a new full TLS session initiated from the proxy server itself to the destination server to retrieve the server certificate details before determining the web category; Once identified, selective inspection can be performed on the real TLS session initiated by the user client.

**With TLS 1.3 ECH**
The entire legacy ClientHello message (Inner ClientHello) is encrypted, encapsulated and sent as part of the new ClientHello wrapper message (Outer ClientHello); So middleboxes cannot identity the destination service anymore and cannot replay the TLS session to the destination server because it has no clue about it. (May be DNS knows but not all the time, see other issue here created by myself).  
So TLS ECH has an impact on enterprise security and compliance (including selective inspection) not because of SNI (which is not reliable) encryption but because there is no other information about the destination server is available and that can be used to fetch and retrieve the server certificate (Which is indeed reliable) to perform web categorization. 
**We don’t care about SNI and its reliability, we need just to know the destination service to get the destination server certificate details.** 

So I think that we need just to adjust our text and explain that SNI is used to retrieve the server certificate details to perform web filtering and selective inspection. With TLS ECH, the ClienHello is encrypted, and the outer SNI does not contain any information about the real destination server. Thus server certificate cannot be retrieved and middleboxes become blind;

**Comment by zied-turki** [comment]
Aligned. Thank you. 
So I note that SNI is not the sole indicator of the intended service. We used other channels or sources of information to confirm the intended web category. 
We should then confirm the SNI reliability and explain that various additional checks are performed by middleboxes – including the server certificate details – to bring this up to a higher level of reliability.  It is not reliable but crucial to a proper functioning of middleboxes.
